src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp

```diff
@@ -552,6 +582,67 @@ class LogicalExpressionOptimizerVisitor : public InDepthQueryTreeVisitorWithCont
         function_node.getArguments().getNodes() = std::move(or_operands);
         function_node.resolveAsFunction(or_function_resolver);
     }
+
+    void tryOptimizeOutRedundantEquals(QueryTreeNodePtr & node)
+    {
+        auto & function_node = node->as<FunctionNode &>();
+        assert(function_node.getFunctionName() == "equals");
+
+        bool lhs_const;
+        bool maybe_invert;
+
+        const ConstantNode * constant;
+        const FunctionNode * child_function;
+
+        const auto function_arguments = function_node.getArguments().getNodes();
+        if (function_arguments.size() != 2)
+            return;
+
+        const auto & lhs = function_arguments[0];
+        const auto & rhs = function_arguments[1];
+
+        if ((constant = lhs->as<ConstantNode>()))
+            lhs_const = true;
+        else if ((constant = rhs->as<ConstantNode>()))
+            lhs_const = false;
+        else
+            return;
+
+        UInt64 val;
+        if (!constant->getValue().tryGet<UInt64>(val))
+            return;
+
+        if (val == 1)
+            maybe_invert = false;
+        else if (val == 0)
+            maybe_invert = true;
+        else
+            return;
+
+        if (lhs_const)
+            child_function = rhs->as<FunctionNode>();
+        else
+            child_function = lhs->as<FunctionNode>();
+
+        if (!child_function || !isBooleanFunction(child_function->getFunctionName()))
+            return;
+
+        if (removeLowCardinality(constant->getResultType())->isNullable())
+            need_rerun_resolve = true;
```

Comment:
Oh, I see. Maybe we can not to consider this case for now and support only `UInt8` constant, because something like `= toNullable(1)` should be pretty rear case, wdyt?